trino-s3-proxy/src/main/java/io/trino/s3/proxy/server/rest/RequestLogger.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.s3.proxy.server.rest;
+
+import com.google.common.base.Stopwatch;
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableMap;
+import io.airlift.log.Logger;
+import io.trino.s3.proxy.spi.signing.SigningServiceType;
+import jakarta.ws.rs.WebApplicationException;
+import org.glassfish.jersey.server.ContainerRequest;
+
+import java.net.URI;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class RequestLogger
+{
+    private static final Logger log = Logger.get(RequestLogger.class);
+
+    private static final LoggerProc debugLogger = new LoggerProc()
+    {
+        @Override
+        public void log(String format, Object... args)
+        {
+            log.debug(format, args);
+        }
+
+        @Override
+        public boolean isEnabled()
+        {
+            return log.isDebugEnabled();
+        }
+    };
+
+    private static final LoggerProc infoLogger = new LoggerProc()
+    {
+        @Override
+        public void log(String format, Object... args)
+        {
+            log.info(format, args);
+        }
+
+        @Override
+        public boolean isEnabled()
+        {
+            return log.isInfoEnabled();
+        }
+    };
+
+    private volatile LoggerProc loggerProc = debugLogger;
+
+    private interface LoggerProc
+    {
+        void log(String format, Object... args);
+
+        boolean isEnabled();
+    }
+
+    // TODO - allow levels to be set for only certain users, IPs, etc.
+
+    public void setLevelInfo()
+    {
+        loggerProc = infoLogger;
+    }
+
+    public void setLevelDebug()
+    {
+        loggerProc = debugLogger;
+    }
+
+    public RequestLoggingSession requestSession(SigningServiceType serviceType, ContainerRequest request)
+    {
+        if (!loggerProc.isEnabled()) {
+            return () -> {};
+        }
+
+        // TODO introduce telemetry/spans
+        UUID requestId = UUID.randomUUID();
+
+        Map<String, String> entries = new ConcurrentHashMap<>();
+
+        Map<String, String> properties = new ConcurrentHashMap<>();
+
+        Map<String, String> errors = new ConcurrentHashMap<>();
+
+        String requestMethod = request.getMethod();
+        URI requestUri = request.getRequestUri();
+        boolean requestHasEntity = request.hasEntity();
+
+        add(entries, "request.id", requestId);
```

Comment:
What do you think about having these as an internal `requestParameters` immutable map? That way we don't need to remember to add them back every time we call `push`

Reply:
Good idea